src/OrleansAzureUtils/Providers/Storage/AzureTableStorage.cs

```diff
@@ -246,49 +261,154 @@ internal void ConvertToStorageFormat(object grainState, GrainStateEntity entity)
 
                 if (Log.IsVerbose3) Log.Verbose3("Writing binary data size = {0} for grain id = Partition={1} / Row={2}",
                     data.Length, entity.PartitionKey, entity.RowKey);
-                
+
                 dataSize = data.Length;
-                entity.Data = data;
+
+                properties = SplitBinaryData(data).Select(t => new EntityProperty(t));
+                basePropertyName = BINARY_DATA_PROPERTY_NAME;
             }
-            if (dataSize > MAX_DATA_SIZE)
+
+            CheckMaxDataSize(dataSize, MAX_DATA_CHUNK_SIZE * MAX_DATA_CHUNKS_COUNT);
+
+            foreach (var keyValuePair in properties.Zip(GetPropertyNames(basePropertyName),
+                (property, name) => new KeyValuePair<string, EntityProperty>(name, property)))
             {
-                var msg = string.Format("Data too large to write to Azure table. Size={0} MaxSize={1}", dataSize, MAX_DATA_SIZE);
+                entity.Properties.Add(keyValuePair);
+            }
+        }
+
+        private void CheckMaxDataSize(int dataSize, int maxDataSize)
+        {
+            if (dataSize > maxDataSize)
+            {
+                var msg = string.Format("Data too large to write to Azure table. Size={0} MaxSize={1}", dataSize, maxDataSize);
                 Log.Error(0, msg);
                 throw new ArgumentOutOfRangeException("GrainState.Size", msg);
             }
         }
 
+        private static IEnumerable<string> SplitStringData(string stringData)
+        {
+            var startIndex = 0;
+            while (startIndex < stringData.Length)
+            {
+                var chunkSize = Math.Min(MAX_STRING_PROPERTY_LENGTH, stringData.Length - startIndex);
+
+                yield return stringData.Substring(startIndex, chunkSize);
+
+                startIndex += chunkSize;
+            }
+        }
+
+        private static IEnumerable<byte[]> SplitBinaryData(byte[] binaryData)
```

Comment:
Regarding previous @DixonD-git comment that was lost in rebase:
> "Not sure how ArraySegments could be used here since we need to split the array physically in order to store it in DynamicTableEntity"

What you have here is fine.
I was suggesting using ArraySegment<byte> due to the following advantages.
- Would not need to allocate any new arrays in the enumeration, making multiple enumerations (always a risk when returning IEnumerable<T>) less costly.
- ArraySegment<byte>.ToArray() would duplicate the array for us, so there is no need to maintain our own array duplication logic (admittedly, not a big savings, but why write/maintain functionality the containers already support).

Loop body would look like:

    {
        var chunkSize = Math.Min(MAX_DATA_CHUNK_SIZE, binaryData.Length - startIndex);  
        yield return new ArraySegment<byte>(binaryData, startIndex, chunkSize)
        startIndex += chunkSize;  
    }

Call would be used like SplitBinaryData(data).Select(t => new EntityProperty(t.ToArray())), with the buffer duplication logic done with ToArray().

Again, what you have is fine.  This is just an elaboration in response to your comment.